src/ape/exceptions.py

```diff
@@ -88,17 +91,26 @@ def __init__(
         base_err: Optional[Exception] = None,
         message: Optional[str] = None,
         code: Optional[int] = None,
+        txn: Optional[TransactionAPI] = None,
     ):
         self.base_err = base_err
         if not message:
             message = str(base_err) if base_err else self.DEFAULT_MESSAGE
 
         self.message = message
         self.code = code
+        self.txn = txn
 
         ex_message = f"({code}) {message}" if code else message
         super().__init__(ex_message)
 
+    @property
+    def trace(self) -> Optional[Iterator["TraceFrame"]]:
+        if self.txn is not None:
+            yield from self.txn.provider.get_transaction_trace(self.txn.txn_hash.hex())
```

Comment:
I believe the existence of a `trace` implies there is a receipt. I am wondering if we should have `receipt` attached to this exception instead of `txn`. This way, we can use the `trace` property on `receipt` (which does the same as this).

Reply:
Yeah I do like that idea more, and it would also simplify a lot of the call sites. Currently since I made it `txn` I ended up just passing `txn` like everywhere that I had one, but if it's switched to `receipt` then I really only have to care about supplying it where it matters (after a transaction invocation).

I'll get this updated, though it will probably mean some changes in the provider plugins since I wrote those in a way that they supply `txn` when raising transaction errors